chia/data_layer/data_store.py

```diff
@@ -350,6 +398,30 @@ async def get_ancestors(self, node_hash: bytes32, tree_id: bytes32, *, lock: boo
 
         return ancestors
 
+    async def get_ancestors_2(self, node_hash: bytes32, tree_id: bytes32, lock: bool = True) -> List[InternalNode]:
+        # TODO: implement with RECURSIVE.
+        nodes = []
+        async with self.db_wrapper.locked_transaction(lock=lock):
```

Comment:
why do we need to lock the database for a read operation? All nodes in our database are immutable, right? There's no chance of data races or nodes disappearing, right?

Reply:
this table is not immutable. suppose we get an update for x -> y, z to x' -> y', z. then z's ancestor changed from x to x'.

Reply:
the update would hash to something other than the previous value, so both entries would exist after the update.
but maybe I'm misunderstanding this table, is this a table of *just* the peak table state?
(if so, I guess I don't understand why we need that, I would expect we could have an index on both key and generation/sequence number)

Reply:
this is just a map from a node to its parent, so we can quickly get the path to the root.